### PR TITLE
[ECO-4986] Implement the `RETRY` room lifecycle operation

### LINKED ISSUE #51: Implement the `RETRY` room lifecycle operation
Split from #28.

- [x] Implement CHA-RL5
- [ ] <del>Implement the part of CHA-RL1h3 and CHA-RL4b9 which triggers a `RETRY`.</del> — will do this in #50

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/ECO-4986) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by lawrence-forooghian**
**Note: This PR is based on top of #115; please review that one first.**

Resolves #51. See commit messages for more details.

## Summary by CodeRabbit

- **New Features**
	- Enhanced contributor management with new retry operation states for improved handling of transient failures.
	- Introduced a structured retry operation method to manage detachment and reattachment of contributors.

- **Bug Fixes**
	- Improved error handling during attachment and detachment processes, ensuring better state management.

- **Tests**
	- Expanded test coverage for retry operations and contributor management, validating the new functionality and error handling improvements.

### CODE REVIEW

File: Sources/AblyChat/RoomLifecycleManager.swift [hunk 1]
```diff
@@ -935,4 +982,94 @@ internal actor DefaultRoomLifecycleManager<Contributor: RoomLifecycleContributor
...
     }
+
+    // MARK: - RETRY operation
+
+    /// Implements CHA-RL5’s RETRY operation.
+    ///
+    /// - Parameters:
+    ///   - forcedOperationID: Allows tests to force the operation to have a given ID. In combination with the ``testsOnly_subscribeToOperationWaitEvents`` API, this allows tests to verify that one test-initiated operation is waiting for another test-initiated operation.
+    ///   - triggeringContributor: This is, in the language of CHA-RL5a, “the channel that became SUSPENDED”.
+    internal func performRetryOperation(testsOnly_forcingOperationID forcedOperationID: UUID? = nil, triggeredByContributor triggeringContributor: Contributor) async {
```
Comment (maratal) [inline Sources/AblyChat/RoomLifecycleManager.swift:993]: Seems like it's not being called from anywhere apart from tests.

Comment (lawrence-forooghian) [reply Sources/AblyChat/RoomLifecycleManager.swift:993]: Correct. See message of f08eabc9b76f5a81b92cdf4c6ef1c59789818005.

Comment (maratal) [reply Sources/AblyChat/RoomLifecycleManager.swift:993]: Oops, sorry 🤦‍♂️

**Review by maratal — APPROVED** [review]

**Review by umair-ably — APPROVED** [review]

**Comment by maratal** [comment]
@lawrence-forooghian feel free to re-request review if changes were significant.

**Comment by lawrence-forooghian** [comment]
@maratal the main change since you reviewed is 8d39600fb0f0a08ee7652c3cb063bcbd5c906e7c, which replaces the previous https://github.com/ably-labs/ably-chat-swift/commit/ec1645ab66a0bce65ba53035206f86ccaaf385d2, splitting the `SUSPENDING` status into two so that I can transition into `SUSPENDED` when the spec says to (i.e. before scheduling the `RETRY`), but still consider there to be an operation in progress once the `RETRY` starts.

**Pull request merged**